src/thor/matrix_action.cc

```diff
@@ -93,33 +63,40 @@ std::string thor_worker_t::matrix(Api& request) {
         case travel_mode_t::kBicycle:
           // Use CostMatrix if number of sources and number of targets
           // exceeds some threshold
-          if (options.sources().size() > kCostMatrixThreshold &&
-              options.targets().size() > kCostMatrixThreshold) {
-            time_distances = costmatrix();
-          } else {
-            time_distances = timedistancematrix();
+          if (options.sources().size() < kCostMatrixThreshold ||
+              options.targets().size() < kCostMatrixThreshold) {
+            matrix_type = MatrixType::TimeDist;
           }
           break;
         case travel_mode_t::kPublicTransit:
-          time_distances = timedistancematrix();
+          matrix_type = MatrixType::TimeDist;
           break;
         default:
-          // force timedistance if traffic is desired and allowed
-          if (has_time(request)) {
-            time_distances = timedistancematrix();
-          } else {
-            time_distances = costmatrix();
-          }
+          break;
       }
       break;
     case COST_MATRIX:
-      time_distances = costmatrix();
+      force_costmatrix = true;
       break;
     case TIME_DISTANCE_MATRIX:
-      time_distances = timedistancematrix();
+      matrix_type = MatrixType::TimeDist;
       break;
   }
-  return tyr::serializeMatrix(request, time_distances, distance_scale);
+
+  // similar to routing: prefer the exact unidirectional algo if not requested otherwise
+  // don't use matrix_type, we only need it to set the right warnings for what will be used
+  bool has_time =
+      check_matrix_time(request,
+                        options.prioritize_bidirectional() ? MatrixType::Cost : MatrixType::TimeDist);
+  if (has_time && !options.prioritize_bidirectional() && !force_costmatrix) {
+    return tyr::serializeMatrix(request, timedistancematrix(), distance_scale, MatrixType::TimeDist);
+  } else if (has_time && options.prioritize_bidirectional()) {
```

Comment:
otherwise, if there's a valid time AND `prioritize_bidirectional` is true, we do a CostMatrix